sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp

```diff
@@ -0,0 +1,317 @@
+__SYCL_JOINT_MATRIX_OVERLOAD_INT8_ARR(b, 8, 32, 4)
+__SYCL_JOINT_MATRIX_OVERLOAD_INT8_ARR(a, 16, 16, 4)
+__SYCL_JOINT_MATRIX_OVERLOAD_INT8_ARR(b, 16, 16, 4)
+
+#undef __SYCL_JOINT_MATRIX_OVERLOAD_INT8_ARR
+
+#define __SYCL_JOINT_MATRIX_OVERLOAD_ARR_ACC(TYPE, M, N)                       \
+  template <>                                                                  \
+  struct joint_matrix_hip<TYPE, matrix_use::accumulator, M, N,                 \
+                          matrix_layout::dynamic> {                            \
+    using ext_array_t =                                                        \
+        __attribute__((__vector_size__((M * N) / 64 * sizeof(TYPE)))) TYPE;    \
+    ext_array_t data = {0};                                                    \
+  };
+
+__SYCL_JOINT_MATRIX_OVERLOAD_ARR_ACC(float, 16, 16)
+__SYCL_JOINT_MATRIX_OVERLOAD_ARR_ACC(float, 32, 32)
+__SYCL_JOINT_MATRIX_OVERLOAD_ARR_ACC(double, 16, 16)
+__SYCL_JOINT_MATRIX_OVERLOAD_ARR_ACC(int32_t, 32, 32)
+__SYCL_JOINT_MATRIX_OVERLOAD_ARR_ACC(int32_t, 16, 16)
+
+#undef __SYCL_JOINT_MATRIX_OVERLOAD_ARR_ACC
+
+template <matrix_layout Layout, typename S, typename T, size_t NumRows,
+          size_t NumCols, access::address_space Space,
+          access::decorated IsDecorated, typename Group>
+void load_accumulator_layoutT(
+    joint_matrix_hip<S, matrix_use::accumulator, NumRows, NumCols,
+                     matrix_layout::dynamic> &res,
+    multi_ptr<T, Space, IsDecorated> src, size_t stride, Group &sg) {
+  std::ignore = stride;
+  const auto idx = sg.get_group_linear_id() * sg.get_local_range()[0] +
+                   sg.get_local_linear_id();
+  if constexpr (Layout == matrix_layout::row_major) {
+    res.data = {0};
+  } else if constexpr (Layout == matrix_layout::col_major) {
+    res.data = {0};
+  }
+};
+
+template <typename Group, typename S, typename T, size_t NumRows,
+          size_t NumCols, access::address_space Space,
+          access::decorated IsDecorated>
+void load_accumulator_hip(
+    joint_matrix_hip<S, matrix_use::accumulator, NumRows, NumCols,
+                     matrix_layout::dynamic> &res,
+    multi_ptr<T, Space, IsDecorated> src, size_t stride, matrix_layout Layout,
+    Group &sg) {
+  switch (Layout) {
+  case matrix_layout::row_major:
+    load_accumulator_layoutT<matrix_layout::row_major>(res, src, stride, sg);
+    break;
+  case matrix_layout::col_major:
+    load_accumulator_layoutT<matrix_layout::col_major>(res, src, stride, sg);
+    break;
+  default:
+    assert(false && "Invalid layout specified!");
+  }
+}
+
+template <
+    typename Group, typename S, typename T, size_t NumRows, size_t NumCols,
+    matrix_use Use, matrix_layout Layout, access::address_space Space,
+    access::decorated IsDecorated,
+    typename = typename std::enable_if_t<(Layout == matrix_layout::row_major ||
+                                          Layout == matrix_layout::col_major)>>
+void load_multiplicand_hip(
```

Comment:
Have you tried loading from const matrix? I'm guessing the type checks will break it below?

Might have to do something similar to https://github.com/intel/llvm/pull/6532/files

Reply:
The type checks have been on `S` which is the type joint_matrix created from. 
Even if `T`(multi_ptr type) is const, it should be fine, as the `const` value is assigned to non-const value of `joint_matrix` data.